File: node/src/generator.js
const VOWELS = new Set(['A', 'E', 'I', 'O', 'U']);

function cleanTokens(name) {
  return (name.toUpperCase().match(/[A-Z0-9]+/g) || []).filter(Boolean);
}

function acronymHints(raw) {
  const hints = raw.match(/\b[A-Z]{2,5}\b/g) || [];
  const seen = new Set();
  const out = [];
  for (const h of hints) {
    if (!seen.has(h)) {
      out.push(h);
      seen.add(h);
    }
  }
  return out;
}

function extractYear(tokens) {
  for (let i = tokens.length - 1; i >= 0; i--) {
    const t = tokens[i];
    if (/^\d{4}$/.test(t)) return t;
  }
  return '';
}

function lettersOnly(tokens) {
  return tokens
    .map((t) => t.replace(/[^A-Z]/g, ''))
    .filter(Boolean);
}

function pronounceabilityScore(code) {
  const letters = code.replace(/[^A-Z]/g, '');
  if (!letters) return 0;
  let v = 0;
  for (const ch of letters) if (VOWELS.has(ch)) v += 1;
  const ratio = v / Math.max(letters.length, 1);
  return 1.0 - Math.min(Math.abs(ratio - 0.45), 0.45) / 0.45;
}

function readabilityScore(code) {
  const badPairs = ['00', '11', 'O0', '0O', 'I1', '1I'];
  let penalty = 0;
  for (const bp of badPairs) if (code.includes(bp)) penalty += 0.2;
  const mix = /[A-Z]/.test(code) && /\d/.test(code);
  return Math.max(0, (mix ? 1.0 : 0.6) - penalty);
}

function lengthScore(code, minLen, maxLen) {
  const ideal = (minLen + maxLen) / 2;
  return Math.max(0, 1.0 - Math.abs(code.length - ideal) / Math.max(ideal, 1));
}

function score(code, minLen, maxLen) {
  return (
    0.45 * pronounceabilityScore(code) +
    0.35 * readabilityScore(code) +
    0.2 * lengthScore(code, minLen, maxLen)
  );
}

function randomDigit() {
  const digits = '23456789';
  return digits[Math.floor(Math.random() * digits.length)];
}

function fit(code, minLen, maxLen) {
  let c = code.toUpperCase().replace(/[^A-Z0-9]/g, '');
  if (c.length > maxLen) c = c.slice(0, maxLen);
  while (c.length < minLen) c += randomDigit();
  return c;
}

function randInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function sample(arr, k) {
  const copy = [...arr];
  const out = [];
  while (copy.length && out.length < k) {
    const idx = randInt(0, copy.length - 1);
    out.push(copy.splice(idx, 1)[0]);
  }
  return out;
}

export function generateCodes(
  campaignName,
  { minLen = 6, maxLen = 12, includeYear = true, count = 8 } = {}
) {
  if (minLen < 6 || maxLen > 12 || minLen > maxLen) {
    throw new Error('length bounds must be within 6-12 and min<=max');
  }

  const raw = campaignName || '';
  const tokens = cleanTokens(raw);
  if (!tokens.length) throw new Error('campaign_name must contain letters or digits');

  const hints = acronymHints(raw);
  const year = includeYear ? extractYear(tokens) : '';
  const year2 = year ? year.slice(-2) : '';
  const words = lettersOnly(tokens);

  const cands = new Set();
  const priority = [];
  for (const a of hints) {
    if (year) priority.push(a + year);
    if (year2) priority.push(a + year2);
    priority.push(a + String(randInt(20, 99)));
  }

  if (words.length) {
    const acronym = words.map((w) => w[0]).join('');
    cands.add(acronym + year);
    cands.add(acronym + year2);

    if (words.length >= 2) {
      cands.add(words[0].slice(0, 2) + words[1].slice(0, 2) + year2);
      cands.add(words[0].slice(0, 3) + words.at(-1).slice(0, 2) + year2);
      cands.add(words[0].slice(0, 2) + words.at(-1).slice(0, 2) + year);
    }

    const main = words[0];
    cands.add(main.slice(0, 4) + year2);
    cands.add(main.slice(0, 3) + (words[1]?.slice(0, 2) || '') + year2);

    if (words.length >= 3) {
      cands.add(words[0].slice(0, 2) + words[1].slice(0, 2) + words[2].slice(0, 2) + year2);
      cands.add(words[0][0] + words[1].slice(0, 2) + words[2].slice(0, 2) + year);
    }
  }

  for (let i = 0; i < 24; i++) {
    const pickCount = Math.min(words.length, [1, 2, 3][randInt(0, 2)]);
    const pick = sample(words, pickCount);
    const frag = pick.map((w) => w.slice(0, [1, 2, 3][randInt(0, 2)])).join('');
    const suffixChoices = [year, year2, String(randInt(20, 99)), String(randInt(2000, 2099))].filter(Boolean);
    const suffix = suffixChoices[randInt(0, suffixChoices.length - 1)] || String(randInt(20, 99));
    cands.add(frag + suffix);
  }

  const normalized = [...cands]
    .map((c) => fit(c, minLen, maxLen))
    .filter((c) => c.length >= minLen && c.length <= maxLen);

  const ranked = [...new Set(normalized)].sort((a, b) => score(b, minLen, maxLen) - score(a, minLen, maxLen));

  const ordered = [];
  const seen = new Set();
  for (const p of priority) {
    const fitted = fit(p, minLen, maxLen);
    if (!seen.has(fitted)) {
      ordered.push(fitted);
      seen.add(fitted);
    }
  }
  for (const c of ranked) {
    if (!seen.has(c)) {
      ordered.push(c);
      seen.add(c);
    }
  }

  return ordered.slice(0, count);
}

File: app/generator.py
import re
import random
from typing import List, Optional

VOWELS = set("AEIOU")


def _clean_tokens(name: str) -> List[str]:
    parts = re.findall(r"[A-Za-z0-9]+", name.upper())
    return [p for p in parts if p]


def _extract_year(tokens: List[str]) -> str:
    for t in reversed(tokens):
        if t.isdigit() and len(t) == 4:
            return t
    return ""


def _letters_only(tokens: List[str]) -> List[str]:
    out = []
    for t in tokens:
        s = re.sub(r"[^A-Z]", "", t)
        if s:
            out.append(s)
    return out


def _pronounceability_score(code: str) -> float:
    letters = ''.join(ch for ch in code if ch.isalpha())
    if not letters:
        return 0
    v = sum(ch in VOWELS for ch in letters)
    ratio = v / max(len(letters), 1)
    # sweet spot around 0.35-0.55
    return 1.0 - min(abs(ratio - 0.45), 0.45) / 0.45


def _readability_score(code: str) -> float:
    bad_pairs = ["00", "11", "O0", "0O", "I1", "1I"]
    penalty = sum(bp in code for bp in bad_pairs) * 0.2
    mix = (any(c.isalpha() for c in code) and any(c.isdigit() for c in code))
    return max(0.0, (1.0 if mix else 0.6) - penalty)


def _length_score(code: str, min_len: int, max_len: int) -> float:
    ideal = (min_len + max_len) / 2
    return max(0.0, 1.0 - abs(len(code) - ideal) / max(ideal, 1))


def _score(code: str, min_len: int, max_len: int) -> float:
    return (
        0.45 * _pronounceability_score(code)
        + 0.35 * _readability_score(code)
        + 0.20 * _length_score(code, min_len, max_len)
    )


def _fit(code: str, min_len: int, max_len: int) -> str:
    code = re.sub(r"[^A-Z0-9]", "", code.upper())
    if len(code) > max_len:
        code = code[:max_len]
    while len(code) < min_len:
        code += random.choice("23456789")
    return code


def generate_codes(
    campaign_name: str,
    min_len: int = 6,
    max_len: int = 12,
    include_year: bool = True,
    count: int = 8,
    seed: Optional[int] = None,
) -> List[str]:
    if seed is not None:
        random.seed(seed)

    if min_len < 6 or max_len > 12 or min_len > max_len:
        raise ValueError("length bounds must be within 6-12 and min<=max")

    tokens = _clean_tokens(campaign_name)
    if not tokens:
        raise ValueError("campaign_name must contain letters or digits")

    year = _extract_year(tokens) if include_year else ""
    year2 = year[-2:] if year else ""
    words = _letters_only(tokens)

    cands = set()

    # patterns
    if words:
        acronym = ''.join(w[0] for w in words if w)
        cands.add(acronym + year)
        cands.add(acronym + year2)

        if len(words) >= 2:
            cands.add(words[0][:2] + words[1][:2] + year2)
            cands.add(words[0][:3] + words[-1][:2] + year2)
            cands.add(words[0][:2] + words[-1][:2] + year)

        main = words[0]
        cands.add(main[:4] + year2)
        cands.add(main[:3] + (words[1][:2] if len(words) > 1 else "") + year2)

        if len(words) >= 3:
            cands.add(words[0][:2] + words[1][:2] + words[2][:2] + year2)
            cands.add(words[0][0] + words[1][:2] + words[2][:2] + year)

    # fallback random blends
    for _ in range(24):
        pick = random.sample(words, k=min(len(words), random.choice([1, 2, 3])))
        frag = ''.join(w[:random.choice([1, 2, 3])] for w in pick)
        suffix = random.choice([year, year2, str(random.randint(20, 99)), str(random.randint(2000, 2099))])
        cands.add(frag + suffix)

    normalized = [_fit(c, min_len, max_len) for c in cands]
    normalized = [c for c in normalized if min_len <= len(c) <= max_len]

    ranked = sorted(set(normalized), key=lambda c: _score(c, min_len, max_len), reverse=True)
    return ranked[:count]
